api/client/client.go

```diff
@@ -0,0 +1,230 @@
+		err := json.Unmarshal([]byte(entity), items[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ModelSave saves existing instances
+func (c *Client) ModelSave(storeID, modelName string, items ...interface{}) error {
+	values, err := marshalItems(items)
+	if err != nil {
+		return err
+	}
+
+	req := &pb.ModelSaveRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		Values:    values,
+	}
+	_, err = c.client.ModelSave(c.ctx, req)
+	return err
+}
+
+// ModelDelete deletes data
+func (c *Client) ModelDelete(storeID, modelName string, entityIDs ...string) error {
+	req := &pb.ModelDeleteRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityIDs: entityIDs,
+	}
+	_, err := c.client.ModelDelete(c.ctx, req)
+	return err
+}
+
+// ModelHas checks if the specified entities exist
+func (c *Client) ModelHas(storeID, modelName string, entityIDs ...string) (bool, error) {
+	req := &pb.ModelHasRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityIDs: entityIDs,
+	}
+	resp, err := c.client.ModelHas(c.ctx, req)
+	if err != nil {
+		return false, err
+	}
+	return resp.GetExists(), nil
+}
+
+// ModelFind finds records by query
+func (c *Client) ModelFind(storeID, modelName string, query es.JSONQuery) ([][]byte, error) {
+	queryBytes, err := json.Marshal(query)
+	if err != nil {
+		return [][]byte{}, err
+	}
+	req := &pb.ModelFindRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		QueryJSON: queryBytes,
+	}
+	resp, err := c.client.ModelFind(c.ctx, req)
+	if err != nil {
+		return [][]byte{}, err
+	}
+	return resp.GetEntities(), nil
+}
+
+// ModelFindByID finds a record by id
+func (c *Client) ModelFindByID(storeID, modelName, entityID string, entity interface{}) error {
+	req := &pb.ModelFindByIDRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityID:  entityID,
+	}
+	resp, err := c.client.ModelFindByID(c.ctx, req)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal([]byte(resp.GetEntity()), entity)
+	return err
+}
+
+// ReadTransaction returns a read transaction that can be started and used and ended
+func (c *Client) ReadTransaction(storeID, modelName string) (*ReadTransaction, error) {
+	client, err := c.client.ReadTransaction(c.ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &ReadTransaction{client: client, storeID: storeID, modelName: modelName}, nil
+}
+
+// WriteTransaction returns a read transaction that can be started and used and ended
+func (c *Client) WriteTransaction(storeID, modelName string) (*WriteTransaction, error) {
+	client, err := c.client.WriteTransaction(c.ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &WriteTransaction{client: client, storeID: storeID, modelName: modelName}, nil
+}
+
+// Listen provides an update whenever the specified model is updated
+func (c *Client) Listen(storeID, modelName, entityID string) (chan []byte, func(), error) {
+	channel := make(chan []byte)
+	ctx, cancel := context.WithCancel(c.ctx)
+	go func() {
```

Comment:
All `err` here are dismissed. 
I see two cases:
- an `err` is returned by `.Recv` since the client canceled the client. So that's a kind of correct-workflow error.
- an `err` because something wrong happened (closed connection, the server died, etc).

I guess some type of checking can be done to detect both cases, and if in the second case:
1. do some `log.Error` to keep track that something bad happened
2. Add some extra documentation in the method like:
```
 // Listen provides an update whenever the specified model is updated
// If the returned channel is closed without the cancel func being called, 
// should be interpreted as an error in the communication-channel with 
// the server.
```
since on any kind of network connection problem, the goroutine finishes, and the returned channel is closed.

Another idiomatic option, in this case, is to return an extra `err chan error`, to allow the client to have visibility that something wrong happened (and actually, what was the error).

Not sure what you think @asutula  @sanderpick .

Reply:
I agree we can improve this. Mind is a little too tired right now to make changes here, so I'll come back to this another day.

Reply:
@jsign what if the returned channel was a channel of `ListenEvent` where:
```
type ListenEvent struct {
    data []byte
    err error
```
That way, we can just send the error on the same channel and close the channel.

Reply:
Yes, that can be a possible option. Check each received `ListenEvent` if `err != nil` and do accordingly.

Reply:
Ok, made that change and addressed the issue of inspecting the error type and just swallowing it silently if it is the error from disposing of the listener.